keras/layers/normalization/batch_normalization.py

```diff
@@ -282,3 +287,57 @@ def get_config(self):
             "synchronized": self.synchronized,
         }
         return {**base_config, **config}
+
+    def _moments(self, inputs, mask):
+        if mask is None:
+            return ops.moments(
+                inputs,
+                axes=self._reduction_axes,
+                synchronized=self.synchronized,
+            )
+
+        mask_weights = ops.cast(
+            mask,
+            inputs.dtype,
+        )
+        mask_weights_broadcasted = ops.expand_dims(
+            mask_weights,
+            axis=-1,
+        )
+        weighted_inputs = mask_weights_broadcasted * inputs
+
+        weighted_input_sum = ops.sum(
+            weighted_inputs,
+            self._reduction_axes,
+            keepdims=True,
+        )
+        sum_of_weights = ops.sum(
+            mask_weights_broadcasted,
+            self._reduction_axes,
+            keepdims=True,
+        )
+        mean = self._divide_no_nan(weighted_input_sum, sum_of_weights)
+
+        difference = inputs - mean
+        squared_difference = difference * difference
+        weighted_distsq = ops.sum(
+            mask_weights_broadcasted * squared_difference,
+            self._reduction_axes,
+            keepdims=True,
+        )
+        variance = self._divide_no_nan(weighted_distsq, sum_of_weights)
+
+        return ops.squeeze(mean), ops.squeeze(variance)
+
+    def _divide_no_nan(self, x1, x2):
```

Comment:
I am not convinced that this is the best solution performance wise (it could be fine, not sure). Would it be better to just divide in float and use an epsilon fuzz factor to avoid division by zero?

Reply:
Yes! Removed in favor of an epsilon approach, good call. The way this is used here mitigates the risks of that approach from negative epsilon values in the denominator (which is generated from the binary masks) and misleadingly large results coming from division by small epsilon (because the numerators are masked).